### ISSUE TITLE
--std=c++20 -fno-modules still sets __has_feature(modules) to 1

### ISSUE THREAD

**Issue by pkasting** [llvm/llvm-project#57432]
The following code should compile without error with `--std=c++20 -fno-modules`:
```
#if __has_feature(modules)
#error Module support found!
#endif
```
Actual result: error message.

**Comment by ChuanqiXu9** [comment]
The `-fmodules` and `-fno-modules` are prepared for clang modules extension instead of the standard c++ modules. This is documented in the preparing document: https://reviews.llvm.org/D131388

**Comment by ChuanqiXu9** [comment]
And it looks true currently we lack the ability to disable the standard c++ modules. May I ask the intention?

**Comment by sdefresne** [comment]
This is in the context of mixing Objective-C++ and Swift.

The Swift compiler generates code that looks like this:

```
#if __has_feature(modules)
@import Foundation;
#endif
```

We are currently building a code base that is C++17 and uses some Swift and Objective-C++ code. Our code base does not yet support building with clang modules support enabled. When trying to build with `-std=c++20`, this `@import` line is reached and causes compilation failure (since we do not pass `-fmodules` and passing it is not enough).

**Comment by ChuanqiXu9** [comment]
Oh, I understood it roughly. Currently both `-std=c++20` and `-fmodules` will enable modules feature too. And it looks like we need to rename them or find a way to disable the cplusplusmodules specially.

**Comment by zmodem** [comment]
It looks like https://reviews.llvm.org/D106864 addressed the command-line flag problem?

**Comment by ChuanqiXu9** [comment]
Yeah, it looks true. Although it requires `-Xclang -fno-cxx-modules`. @pkasting could you give a try?

**Comment by rnk** [comment]
In the long run, we would like to avoid using Xclang flags in the Chromium build. Is there a plan to make this a driver flag?

**Comment by ChuanqiXu9** [comment]
Yeah, there is an issue tracking this: https://github.com/llvm/llvm-project/issues/55891. Although we don't have any concrete proposal yet...
